File: typescript/src/storage.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { Task } from "./task.js";
import { StorageError } from "./errors.js";

export function saveTasks(filePath: string, tasks: Task[]): void {
  try {
    mkdirSync(dirname(filePath), { recursive: true });
    const data = tasks.map((t) => t.toDict());
    writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
  } catch (e) {
    throw new StorageError(`Impossible d'ecrire dans ${filePath} : ${(e as Error).message}`);
  }
}

export function loadTasks(filePath: string): Task[] {
  if (!existsSync(filePath)) return [];
  try {
    const raw = readFileSync(filePath, "utf-8");
    if (!raw.trim()) return [];
    const data = JSON.parse(raw);
    return (data as any[]).map((d) => Task.fromDict(d));
  } catch (e) {
    throw new StorageError(`Lecture impossible de ${filePath} : ${(e as Error).message}`);
  }
}

File: python/src/storage.py
"""Persistance JSON des taches."""

import json
from pathlib import Path
from .task import Task
from .exceptions import StorageError


def save_tasks(file_path, tasks):
    try:
        path = Path(file_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        data = [task.to_dict() for task in tasks]
        path.write_text(json.dumps(data, indent=2, ensure_ascii=False), encoding="utf-8")
    except OSError as e:
        raise StorageError(f"Impossible d'ecrire dans {file_path} : {e}")


def load_tasks(file_path):
    path = Path(file_path)
    if not path.exists():
        return []
    try:
        raw = path.read_text(encoding="utf-8")
        if not raw.strip():
            return []
        data = json.loads(raw)
        return [Task.from_dict(d) for d in data]
    except json.JSONDecodeError as e:
        raise StorageError(f"Fichier JSON corrompu ({file_path}) : {e}")
    except OSError as e:
        raise StorageError(f"Impossible de lire {file_path} : {e}")
